api/v1beta1/cloudstackmachine_webhook.go

```diff
@@ -55,18 +57,25 @@ var _ webhook.Validator = &CloudStackMachine{}
 func (r *CloudStackMachine) ValidateCreate() error {
 	cloudstackmachinelog.Info("validate create", "name", r.Name)
 
-	var (
-		errorList field.ErrorList
-		spec      = r.Spec
-	)
+	var errorList field.ErrorList
 
 	// IdentityRefs must be Secrets.
-	if spec.IdentityRef != nil && spec.IdentityRef.Kind != defaultIdentityRefKind {
+	if r.Spec.IdentityRef != nil && r.Spec.IdentityRef.Kind != defaultIdentityRefKind {
 		errorList = append(errorList, field.Forbidden(field.NewPath("spec", "identityRef", "kind"), "must be a Secret"))
 	}
 
-	errorList = webhook_utilities.EnsureFieldExists(spec.Offering, "Offering", errorList)
-	errorList = webhook_utilities.EnsureFieldExists(spec.Template, "Template", errorList)
+	affinity := strings.ToLower(r.Spec.Affinity)
+	if !(affinity == "" || affinity == "no" || affinity == "pro" || affinity == "anti") {
+		errorList = append(errorList, field.Invalid(field.NewPath("spec", "Affinity"), r.Spec.Affinity,
+			`Affinity must be "no", "pro", "anti", or unspecified.`))
+	}
+	if (affinity != "no") && affinity != "" && (len(r.Spec.AffinityGroupIds) > 0) {
```

Comment:
some inconsistency with the parentheses here which would be good to clean up. Alternately, can we specify these attributes as mutually exclusive, as in https://github.com/go-playground/validator/issues/608#issuecomment-956253612?

Reply:
Cleaned up parens. I think your suggested method is probably a better way to do this, but I don't want to put it through just yet. I think there may be issues with `no` and `""` both being considered unspecified.